tests/robustness/traffic/kubernetes.go

```diff
@@ -61,58 +65,97 @@ const (
 )
 
 func (t kubernetesTraffic) Run(ctx context.Context, clientId int, c *RecordingClient, limiter *rate.Limiter, ids identity.Provider, lm identity.LeaseIdStorage, finish <-chan struct{}) {
-	for {
-		select {
-		case <-ctx.Done():
-			return
-		case <-finish:
-			return
-		default:
-		}
-		objects, err := t.Range(ctx, c, "/registry/"+t.resource+"/", true)
-		if err != nil {
-			continue
+	s := newStorage()
+	keyPrefix := "/registry/" + t.resource + "/"
+	g := errgroup.Group{}
+
+	g.Go(func() error {
+		for {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-finish:
+				return nil
+			default:
+			}
+			resp, err := t.Range(ctx, c, keyPrefix, true)
+			if err != nil {
+				continue
+			}
+			s.Reset(resp)
+			limiter.Wait(ctx)
+			watchCtx, cancel := context.WithTimeout(ctx, WatchTimeout)
+			for e := range c.Watch(watchCtx, keyPrefix, resp.Header.Revision, true) {
+				s.Update(e)
+			}
+			cancel()
 		}
-		limiter.Wait(ctx)
-		err = t.Write(ctx, c, ids, objects)
-		if err != nil {
-			continue
+	})
+	g.Go(func() error {
+		lastWriteFailed := false
+		for {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-finish:
+				return nil
+			default:
+			}
+			// Avoid multiple failed writes in a row
+			if lastWriteFailed {
+				resp, err := t.Range(ctx, c, keyPrefix, true)
+				if err != nil {
+					continue
+				}
+				s.Reset(resp)
+				limiter.Wait(ctx)
+			}
+			err := t.Write(ctx, c, ids, s)
+			lastWriteFailed = err != nil
+			if err != nil {
+				continue
+			}
```

Comment:
```suggestion
			if err := t.Write(ctx, c, ids, s); err != nil {
				lastWriteFailed = true
				continue
			}
```

Reply:
minor comment: feel free to keep it as it's

Reply:
Prefer the old version, reason is that it better shows intention and it's less prone to code changes.